dgl_sparse/src/cpu/matrix_ops_impl.cc

```diff
@@ -0,0 +1,69 @@
+/**
+ *  Copyright (c) 2023 by Contributors
+ * @file cpu/matrix_ops_impl.cc
+ * @brief DGL CPU implementation of matrix operators.
+ */
+#include "../matrix_ops_impl.h"
+
+#include "../utils.h"
+
+namespace dgl {
+namespace sparse {
+
+template <c10::DeviceType XPU, typename IdType>
+std::tuple<std::shared_ptr<COO>, torch::Tensor, torch::Tensor>
+COOIntersectionImpl(
+    const std::shared_ptr<COO> &lhs, const std::shared_ptr<COO> &rhs) {
+  const int64_t lhs_len = lhs->indices.size(1);
+  const int64_t rhs_len = rhs->indices.size(1);
+
+  torch::Tensor lhs_row = lhs->indices.index({0}).contiguous();
+  const IdType *lhs_row_ptr = static_cast<IdType *>(lhs_row.data_ptr());
+  torch::Tensor lhs_col = lhs->indices.index({1}).contiguous();
+  const IdType *lhs_col_ptr = static_cast<IdType *>(lhs_col.data_ptr());
+  torch::Tensor rhs_row = rhs->indices.index({0}).contiguous();
+  const IdType *rhs_row_ptr = static_cast<IdType *>(rhs_row.data_ptr());
+  torch::Tensor rhs_col = rhs->indices.index({1}).contiguous();
+  const IdType *rhs_col_ptr = static_cast<IdType *>(rhs_col.data_ptr());
+
+  std::vector<IdType> ret_rows, ret_cols;
+  std::vector<IdType> ret_lhs_indices, ret_rhs_indices;
+
+  auto pair_hash_fn = [](const std::pair<IdType, IdType> &pair) {
+    return std::hash<IdType>()(pair.first) ^ std::hash<IdType>()(pair.second);
+  };
+  std::unordered_map<std::pair<IdType, IdType>, IdType, decltype(pair_hash_fn)>
+      pair_map(lhs_len, pair_hash_fn);
+  for (int64_t k = 0; k < lhs_len; ++k) {
+    pair_map.emplace(std::make_pair(lhs_row_ptr[k], lhs_col_ptr[k]), k);
+  }
```

Comment:
If we don't consider operator efficiency. We can simply do the following (inspired by https://discuss.pytorch.org/t/intersection-between-to-vectors-tensors/50364/10?u=minjie_wang)

1. Hash the integer pairs to one integer (e.g., by `i * N + j` where `N` is the number of rows/columns).
2. Concatenate lhs and rhs into one vector.
3. Compute unique and the unique indices.
4. Unhash the integer to integer pairs and recover the indices.

This way you may get an operator that supports both CPU/GPU.

Reply:
It is doable and not very slow. I think we can use this implementation. Then do you think we need to keep the current code structure for adding a new operator?